untyped-lambda-calculus/article.md

```diff
@@ -0,0 +1,492 @@
+>
+> -- James Iry, [A Brief, Incomplete, and Mostly Wrong History of Programming Languages][brief-incomplete-history]
+
+[brief-incomplete-history]: http://james-iry.blogspot.com/2009/05/brief-incomplete-and-mostly-wrong.html
+
+Lambda calculus, initially envisioned as a formal logic system, was developed by Alonzo Church around the 1930s to explore the foundations of mathematics.
+The initial formulation had a logical inconsistency known as [Kleene–Rosser paradox] (for a more comprehensive overview, refer to [Paradoxes and Contemporary Logic][paradoxes-contemporary-logic]).
+To sidestep this issue, Church isolated the part of lambda calculus relevant only to computation in 1936.
+This isolate is now known as the untyped lambda calculus.
+
+Later, in 1940, a typed version of lambda calculus based on Russel's type theory was introduced.
+It has weaker expressive power but it's logically consistent.
+
+[Kleene–Rosser paradox]: https://en.wikipedia.org/wiki/Kleene%E2%80%93Rosser_paradox
+[paradoxes-contemporary-logic]: https://plato.stanford.edu/entries/paradoxes-contemporary-logic/#IncoCertFormLogi
+
+In the mid-1960s, Peter Landin showed that lambda calculus models arbitrarily complex programming languages.
+This insight arguably kickstarted the research on functional programming languages.
+
+## The pure untyped lambda calculus
+
+The simplest, smallest type of lambda calculus is the pure untyped lambda calculus, so we'll start with it.
+
+The untyped lambda calculus is, first and foremost, a consistent system for rewriting expressions.
+Lambda calculus that only contains functions and variables is called pure, meaning it doesn't have anything beyond what's strictly necessary.
+
+We can formally define the pure untyped lambda calculus in terms of _abstractions_ and _applications_.
+
+### Abstractions
+
+An _abstraction_, or a functional abstraction, is a parametric expression, that is to say, a function.
+Lambda calculus only defines univariate (single-variable) functions, but we can easily extend it to include multivariate functions too, as we'll see later.
+
+We introduce new abstractions using the $\lambda$ symbol.
+An abstraction consists of a _head_ and a _body_, separated by the dot ($.$) symbol.
+The head contains the $\lambda$ symbol and the parameter name.
+The body is an arbitrary expression.
+For example, $\lambda x. x$ is an abstraction.
+
+As usual in abstract mathematics, we can assign a name to an expression.
+Here we'll be using the equals sign ($=$) for that.
+For example: $id = \lambda x. x.$
+
+Lambda calculus derives its name from the $\lambda$ symbol used in this notation.
+The symbol was most likely chosen arbitrarily.
+However, there's an apocryphal story the notation was derived from $\hat x$ morphing first into $\Lambda x$ and later into $\lambda x$.
+
+Variables in the body of an abstraction that are parameters of said abstraction are called _bound variables_.
+Other variables are called _free variables_.
+
```

Comment:
Can we have some examples here?

Reply:
Added a couple examples. Sorry if the phrasing is a bit awkward, I'm feeling a little under the weather today, so words come a little harder than usual.